### PR TITLE
Added the ability to give permissions on the campaigns.

### PR DESCRIPTION
**Pull request by Noa83**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | No
| New feature? | Enhancement
| Automated tests included? | No
| Related user documentation PR URL | None
| Related developer documentation PR URL | None
| Issues addressed (#s or URLs) | [#6166](https://github.com/mautic/mautic/issues/6166)
| BC breaks? | No
| Deprecations? | No

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR gives the alternative to set permissions in the roles concerning campaigns.

#### Steps to test this PR:
1. Apply the PR
2. As an administrator, create a role with only view own/edit own/create/delete own/publish own selected for the campaigns permissions and the necessary permissions on others categories in order to build a campaign.
3. Create 2 new users with this role.
4. Create a campaign as an administrator
5. Logout and login with one of the new user.
6. Go to the campaign tab and note that you don't see the previously created campaign (whose belongs to admin).
7. Create a campaign. 
8. Logout and login with the second user. Then, repeat the step 6. You don't be able to see neither the admin's campaign nor the first user's campaign.
9. Each action of delete/edit/clone are covered by the permissions so you can't delete/edit/clone the campaign of anyone else when you're logged in as one or the other users you've created, even if you try to change it by the url way.

NB: Testing this PR have permitted to find a bug concerning the clone of some objects when the "own" permissions were set. This bug is fixed in this PR [#6238](https://github.com/mautic/mautic/pull/6238). So the permissions enhancement is complete for the campaigns.

### CODE REVIEW

**Comment by escopecz** [comment]
Travis is reporting lots of failed tests. Please investigate.

File: app/bundles/CampaignBundle/Entity/CampaignRepository.php [hunk 1]
```diff
@@ -103,6 +104,11 @@ public function getPublishedCampaigns($specificId = null, $leadId = null, $forLi
         $q->leftJoin('c.lists', 'll')
             ->where($this->getPublishedByDateExpression($q));
 
+        if (!$viewOther) {
+            $q->andWhere($q->expr()->eq('c.createdBy', ':id'))
```
Comment (Woeler) [inline app/bundles/CampaignBundle/Entity/CampaignRepository.php]: Column name is `created_by` not `createdBy`.

File: app/bundles/CampaignBundle/Entity/CampaignRepository.php [hunk 2]
```diff
@@ -150,6 +156,12 @@ public function getPublishedCampaignsByLeadLists($leadLists)
         $q->andWhere(
             $q->expr()->in('ll.leadlist_id', $leadLists)
         );
+
+        if (!$viewOther) {
+            $q->andWhere($q->expr()->eq('c.createdBy', ':id'))
```
Comment (Woeler) [inline app/bundles/CampaignBundle/Entity/CampaignRepository.php]: Column name is `created_by` not `createdBy`.

**Comment by Noa83** [comment]
@Woeler,  I've made the changes you requested.

**Comment by escopecz** [comment]
@npracht did you test this as you added the Pending Test Confirmation label?

**Comment by escopecz** [comment]
I will assume that Norman tested this PR as well so this could get into 2.15.0 today.

**Comment by npracht** [comment]
Oh yes sorry, i missed the review feature of GH -_-

**Comment by alanhartless** [comment]
@npracht @Woeler did we test this for BC breaks with current permissions?

**Comment by npracht** [comment]
@alanhartless do you have a case to propose i should especially care about ?

**Comment by alanhartless** [comment]
Configure Mautic with a role that has campaign permissions _before_ applying the PR. Make sure they have full campaign permissions. 

Add the PR and go to the permission. What is selected by default? 

Login as the user with the restricted role. Are they allowed to view and edit pre PR campaigns?

**Comment by Enc3phale** [comment]
If `full` is checked before PR, `full` is checked after.

But if this are checked before PR:
screenshot_2018-12-04 roles - edit webmaster mautic
After PR it's like this:
screenshot_2018-12-04 roles - edit webmaster mautic 1

**Comment by npracht** [comment]
Ok for `full` it seems clean. For specified permissions I don't find it very logic. It should be minimum access to own, and eventually to others, but not only others... The best would be own and others as far as before it was not distinguished.
@alanhartless do you have any idea how we could address that ?

**Comment by Enc3phale** [comment]
If I save nothing change, "own" are not checked. And I can't see campaigns with this account.

**Comment by Woeler** [comment]
Will you guys be able to provide a fix for that before the day ends? Otherwise I will have to take it out of the dev release.

**Comment by npracht** [comment]
Working on it, not easy with the serialized value. If you have any tip of example we can save some time :D

**Comment by Enc3phale** [comment]
It's done, could you check code please ?
https://github.com/mautic/mautic/pull/6963
Thanks!

**Pull request merged**